# 📊 Telecom Customer Churn — Exploratory Data Analysis

**Dataset:** `telecom_churn.csv`  
**Goal:** Understand the structure of the dataset and explore patterns that may explain why customers churn.

---

## 📋 Table of Contents
1. [Import Libraries](#1)
2. [Load & Preview Data](#2)
3. [Data Types & Structure](#3)
4. [Descriptive Statistics](#4)
5. [Target Variable Analysis (Churn)](#5)
6. [Sorting & Indexing](#6)
7. [Filtering & Conditional Queries](#7)
8. [Applying Functions](#8)
9. [Encoding Categorical Variables](#9)
10. [GroupBy Analysis](#10)
11. [Summary Tables — Crosstab & Pivot](#11)
12. [Feature Engineering](#12)

---
## 1. Import Libraries <a id='1'></a>

In [1]:
import numpy as np
import pandas as pd

---
## 2. Load & Preview Data <a id='2'></a>

We load the telecom churn dataset and preview the first few rows to understand its structure.

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/Shontu72/python_data-science-learning__-/main/DATA/telecom_churn.csv')
df.head()

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
0,KS,128,415,No,Yes,25,265.1,110,45.07,197.4,99,16.78,244.7,91,11.01,10.0,3,2.70,1,False
1,OH,107,415,No,Yes,26,161.6,123,27.47,195.5,103,16.62,254.4,103,11.45,13.7,3,3.70,1,False
2,NJ,137,415,No,No,0,243.4,114,41.38,121.2,110,10.30,162.6,104,7.32,12.2,5,3.29,0,False
3,OH,84,408,Yes,No,0,299.4,71,50.90,61.9,88,5.26,196.9,89,8.86,6.6,7,1.78,2,False
4,OK,75,415,Yes,No,0,166.7,113,28.34,148.3,122,12.61,186.9,121,8.41,10.1,3,2.73,3,False


**Dataset shape** — how many rows and columns we are working with.

In [3]:
print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")
df.columns.tolist()

Rows: 3333, Columns: 20


['State',
 'Account length',
 'Area code',
 'International plan',
 'Voice mail plan',
 'Number vmail messages',
 'Total day minutes',
 'Total day calls',
 'Total day charge',
 'Total eve minutes',
 'Total eve calls',
 'Total eve charge',
 'Total night minutes',
 'Total night calls',
 'Total night charge',
 'Total intl minutes',
 'Total intl calls',
 'Total intl charge',
 'Customer service calls',
 'Churn']

---
## 3. Data Types & Structure <a id='3'></a>

We check data types and verify there are no missing values.

In [4]:
# Summary of dtypes, non-null counts, and memory usage
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

In [5]:
# Check for missing values per column
df.isnull().sum()

,0
State,0
Account length,0
Area code,0
International plan,0
Voice mail plan,0
Number vmail messages,0
Total day minutes,0
Total day calls,0
Total day charge,0
Total eve minutes,0


> ✅ **No null values** in this dataset — no imputation needed.

Now let's identify which columns are categorical (object dtype).

In [6]:
# Inspect object-type (categorical) columns
df.select_dtypes(include='object').head()

,State,International plan,Voice mail plan
0,KS,No,Yes
1,OH,No,Yes
2,NJ,No,No
3,OH,Yes,No
4,OK,Yes,No


**Categorical columns identified:** `State`, `International plan`, `Voice mail plan`.

---
## 4. Descriptive Statistics <a id='4'></a>

We convert `Churn` from boolean to integer (0/1) so it's included in `.describe()` and arithmetic operations.

In [7]:
# Convert boolean Churn to integer for numeric analysis
df['Churn'] = df['Churn'].astype('int64')

# Numeric summary statistics
df.describe()

,Account length,Area code,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
count,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000,3333.000000
mean,101.064806,437.182418,8.099010,179.775098,100.435644,30.562307,200.980348,100.114311,17.083540,200.872037,100.107711,9.039325,10.237294,4.479448,2.764581,1.562856,0.144914
std,39.822106,42.371290,13.688365,54.467389,20.069084,9.259435,50.713844,19.922625,4.310668,50.573847,19.568609,2.275873,2.791840,2.461214,0.753773,1.315491,0.352067
min,1.000000,408.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,23.200000,33.000000,1.040000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,74.000000,408.000000,0.000000,143.700000,87.000000,24.430000,166.600000,87.000000,14.160000,167.000000,87.000000,7.520000,8.500000,3.000000,2.300000,1.000000,0.000000
50%,101.000000,415.000000,0.000000,179.400000,101.000000,30.500000,201.400000,100.000000,17.120000,201.200000,100.000000,9.050000,10.300000,4.000000,2.780000,1.000000,0.000000
75%,127.000000,510.000000,20.000000,216.400000,114.000000,36.790000,235.300000,114.000000,20.000000,235.300000,113.000000,10.590000,12.100000,6.000000,3.270000,2.000000,0.000000
max,243.000000,510.000000,51.000000,350.800000,165.000000,59.640000,363.700000,170.000000,30.910000,395.000000,175.000000,17.770000,20.000000,20.000000,5.400000,9.000000,1.000000


In [8]:
# Summary for categorical columns
# (After encoding below, this will be less relevant — run before encoding)
df.describe(include=['object'])

,State,International plan,Voice mail plan
count,3333,3333,3333
unique,51,2,2
top,WV,No,No
freq,106,3010,2411


**Observations:**
- Most common state: `WV` (106 occurrences)
- `International plan`: 3010 customers have `No` (majority)
- `Voice mail plan`: 2411 customers have `No`

---
## 5. Target Variable Analysis (Churn) <a id='5'></a>

Churn is our target variable. It follows a **Bernoulli distribution** — its mean equals P(Churn = 1).

In [9]:
# Absolute counts
print("Churn value counts:")
print(df['Churn'].value_counts())

Churn value counts:
Churn
0    2850
1     483
Name: count, dtype: int64


In [10]:
# Proportions — the mean of a binary column equals P(1)
print("Churn proportions (Bernoulli distribution):")
print(df['Churn'].value_counts(normalize=True))

print(f"\nChurn rate: {df['Churn'].mean():.2%}")

Churn proportions (Bernoulli distribution):
Churn
0    0.855086
1    0.144914
Name: proportion, dtype: float64

Churn rate: 14.49%


> ⚠️ **~14.5% churn rate** — This is quite high. A churn rate at this level can severely impact revenue and is a strong business signal to investigate further.

This also means the dataset is **imbalanced** (85% non-churn vs 15% churn), which is important to keep in mind for any future ML modeling.

---
## 6. Sorting & Indexing <a id='6'></a>

Useful for quickly finding extreme values — e.g., who pays the most?

In [11]:
# Top 5 customers by daytime charge (highest bill first)
df.sort_values(by='Total day charge', ascending=False).head()

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
365,CO,154,415,No,No,0,350.8,75,59.64,216.5,94,18.40,253.9,100,11.43,10.1,9,2.73,1,1
985,NY,64,415,Yes,No,0,346.8,55,58.96,249.5,79,21.21,275.4,102,12.39,13.3,9,3.59,1,1
2594,OH,115,510,Yes,No,0,345.3,81,58.70,203.4,106,17.29,217.5,107,9.79,11.8,8,3.19,1,1
156,OH,83,415,No,No,0,337.4,120,57.36,227.4,116,19.33,153.9,114,6.93,15.8,7,4.27,0,1
605,MO,112,415,No,No,0,335.5,77,57.04,212.5,109,18.06,265.0,132,11.93,12.7,8,3.43,2,1


In [12]:
# Find the column index of 'Total day charge'
print(f"'Total day charge' is column index: {df.columns.get_loc('Total day charge')}")

'Total day charge' is column index: 8


In [13]:
# Multi-column sort: first by Churn (ascending), then by Total day charge (descending)
# This shows the highest-paying loyal customers first, then highest-paying churned customers
df.sort_values(
    by=['Churn', 'Total day charge'],
    ascending=[True, False]
).head()

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
688,MN,13,510,No,Yes,21,315.6,105,53.65,208.9,71,17.76,260.1,123,11.70,12.1,3,3.27,3,0
2259,NC,210,415,No,Yes,31,313.8,87,53.35,147.7,103,12.55,192.7,97,8.67,10.1,7,2.73,3,0
534,LA,67,510,No,No,0,310.4,97,52.77,66.5,123,5.65,246.5,99,11.09,9.2,10,2.48,4,0
575,SD,114,415,No,Yes,36,309.9,90,52.68,200.3,89,17.03,183.5,105,8.26,14.2,2,3.83,1,0
2858,AL,141,510,No,Yes,28,308.0,123,52.36,247.8,128,21.06,152.9,103,6.88,7.4,3,2.00,1,0


In [14]:
# loc: label-based slicing — rows 0–5, columns State to Area code
df.loc[0:5, 'State':'Area code']

,State,Account length,Area code
0,KS,128,415
1,OH,107,415
2,NJ,137,415
3,OH,84,408
4,OK,75,415
5,AL,118,510


---
## 7. Filtering & Conditional Queries <a id='7'></a>

Filtering allows us to answer specific business questions.

In [15]:
# Q1: What is the average daytime usage for churned customers?
avg_day_min_churned = df[df['Churn'] == 1]['Total day minutes'].mean()
print(f"Avg day minutes (churned): {avg_day_min_churned:.2f}")

Avg day minutes (churned): 206.91


In [16]:
# Q2: What is the max international call duration for loyal customers WITHOUT an international plan?
# Note: requires parentheses around each condition when using & or |
max_intl = df[(df['Churn'] == 0) & (df['International plan'] == 'No')]['Total intl minutes'].max()
print(f"Max intl minutes (loyal, no intl plan): {max_intl}")

Max intl minutes (loyal, no intl plan): 18.9


In [17]:
# Q3: Mean numeric features for churned users
df[df['Churn'] == 1].select_dtypes(include=np.number).mean()

,0
Account length,102.664596
Area code,437.817805
Number vmail messages,5.115942
Total day minutes,206.914079
Total day calls,101.335404
Total day charge,35.175921
Total eve minutes,212.410145
Total eve calls,100.561077
Total eve charge,18.054969
Total night minutes,205.231677


---
## 8. Applying Functions <a id='8'></a>

`apply()` lets us run any function across rows or columns.

In [18]:
# Apply np.max across all columns — returns the max value per column
df.apply(np.max)

,0
State,WY
Account length,243
Area code,510
International plan,Yes
Voice mail plan,Yes
Number vmail messages,51
Total day minutes,350.8
Total day calls,165
Total day charge,59.64
Total eve minutes,363.7


In [19]:
# Lambda function: filter rows where State starts with 'W'
# df['State'].apply(lambda x: x[0] == 'W') returns a boolean Series
# use it to filter the full DataFrame
df[df['State'].apply(lambda x: x[0] == 'W')].head()

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
9,WV,141,415,Yes,Yes,37,258.6,84,43.96,222.0,111,18.87,326.4,97,14.69,11.2,5,3.02,0,0
26,WY,57,408,No,Yes,39,213.0,115,36.21,191.1,112,16.24,182.7,115,8.22,9.5,3,2.57,0,0
44,WI,64,510,No,No,0,154.0,67,26.18,225.8,118,19.19,265.3,86,11.94,3.5,3,0.95,1,0
49,WY,97,415,No,Yes,24,133.2,135,22.64,217.2,58,18.46,70.6,79,3.18,11.0,3,2.97,1,0
54,WY,87,415,No,No,0,151.0,83,25.67,219.7,116,18.67,203.9,127,9.18,9.7,3,2.62,5,1


---
## 9. Encoding Categorical Variables <a id='9'></a>

We convert Yes/No columns to boolean (True/False) using two approaches:
- `map()` — for a single column
- `replace()` — for applying the same mapping across multiple columns at once

In [20]:
# Mapping dictionary: 'No' -> False, 'Yes': True
d = {'No': False, 'Yes': True}

# Use replace() for both columns for robustness against re-execution
df.replace({'International plan': d, 'Voice mail plan': d}, inplace=True)

df.head()

/tmp/ipykernel_9984/3542410418.py:5: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df.replace({'International plan': d, 'Voice mail plan': d}, inplace=True)


,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,Total eve calls,Total eve charge,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn
0,KS,128,415,False,True,25,265.1,110,45.07,197.4,99,16.78,244.7,91,11.01,10.0,3,2.70,1,0
1,OH,107,415,False,True,26,161.6,123,27.47,195.5,103,16.62,254.4,103,11.45,13.7,3,3.70,1,0
2,NJ,137,415,False,False,0,243.4,114,41.38,121.2,110,10.30,162.6,104,7.32,12.2,5,3.29,0,0
3,OH,84,408,True,False,0,299.4,71,50.90,61.9,88,5.26,196.9,89,8.86,6.6,7,1.78,2,0
4,OK,75,415,True,False,0,166.7,113,28.34,148.3,122,12.61,186.9,121,8.41,10.1,3,2.73,3,0


---
## 10. GroupBy Analysis <a id='10'></a>

GroupBy lets us compare subgroups. Syntax: `df.groupby(by=[cols])[cols_to_show].function()`

In [21]:
# Compare call duration stats between churned vs non-churned customers
# percentiles=[] omits quartiles for a cleaner output
df.groupby(['Churn'])[
    ['Total day minutes', 'Total eve minutes', 'Total night minutes']
].describe(percentiles=[])

Total day minutes                                            \
                  count        mean        std  min    50%    max   
Churn                                                               
0                2850.0  175.175754  50.181655  0.0  177.2  315.6   
1                 483.0  206.914079  68.997792  0.0  217.6  350.8   

      Total eve minutes                                             \
                  count        mean        std   min    50%    max   
Churn                                                                
0                2850.0  199.043298  50.292175   0.0  199.6  361.8   
1                 483.0  212.410145  51.728910  70.9  211.3  363.7   

      Total night minutes                                              
                    count        mean        std   min     50%    max  
Churn                                                                  
0                  2850.0  200.133193  51.105032  23.2  200.25  395.0  
1                   483.0  205.231677  47.132825  47.4  204.80  354.9

In [22]:
# Compare call duration by International plan status
df.groupby(['International plan'])[
    ['Total day minutes', 'Total eve minutes', 'Total night minutes']
].describe(percentiles=[])

Total day minutes                                      \
                               count        mean        std   min    50%   
International plan                                                         
False                         3010.0  178.893887  54.141286   0.0  179.0   
True                           323.0  187.986997  56.850451  12.5  188.9   

                          Total eve minutes                               \
                      max             count        mean        std   min   
International plan                                                         
False               350.8            3010.0  200.663090  50.474772   0.0   
True                346.8             323.0  203.936842  52.879435  60.8   

                                  Total night minutes                         \
                       50%    max               count        mean        std   
International plan                                                             
False               201.35  361.8              3010.0  201.350831  50.366837   
True                202.90  363.7               323.0  196.410217  52.333632   

                                         
                     min     50%    max  
International plan                       
False               23.2  201.85  395.0  
True                72.4  196.20  352.5

In [23]:
# How many churned/loyal customers have an international plan? (absolute counts)
df.groupby(['Churn'])['International plan'].value_counts()

Churn  International plan
0      False                 2664
       True                   186
1      False                  346
       True                   137
Name: count, dtype: int64

In [24]:
# Same, but as proportions within each Churn group
# Key insight: churned customers are ~4x more likely to have an international plan
df.groupby(['Churn'])['International plan'].value_counts(normalize=True)

Churn  International plan
0      False                 0.934737
       True                  0.065263
1      False                 0.716356
       True                  0.283644
Name: proportion, dtype: float64

In [25]:
# Detailed look: churned customers WITH an international plan
df[
    (df['Churn'] == 1) & (df['International plan'] == True)
][['Total day minutes', 'Total eve minutes', 'Total night minutes']].describe(percentiles=[])

,Total day minutes,Total eve minutes,Total night minutes
count,137.000000,137.000000,137.000000
mean,193.625547,209.305839,194.640146
std,61.511100,53.886272,49.330174
min,46.500000,93.700000,72.400000
50%,195.500000,206.300000,192.200000
max,346.800000,363.700000,332.700000


---
## 11. Summary Tables — Crosstab & Pivot <a id='11'></a>

### 11a. Crosstab
Quick frequency table between two categorical variables.

In [26]:
# Absolute counts: Churn vs International plan-- same work but much more easire than previous groupby function
pd.crosstab(df['Churn'], df['International plan'])

International plan,False,True
Churn,,
0,2664,186
1,346,137


In [27]:
# Proportions: Churn vs Voice mail plan
# normalize=True divides every cell by the total count
pd.crosstab(df['Churn'], df['Voice mail plan'], normalize=True)

Voice mail plan,False,True
Churn,,
0,0.602460,0.252625
1,0.120912,0.024002


### 11b. Pivot Table
More flexible than crosstab — lets you choose aggregation function and which values to show.

**Syntax:** `df.pivot_table([values], [index], aggfunc='mean')`

In [28]:
# Average number of calls (day/eve/night) broken down by Area code
df.pivot_table(
    ['Total day calls', 'Total eve calls', 'Total night calls'],
    ['Area code'],
    aggfunc='mean'
)

,Total day calls,Total eve calls,Total night calls
Area code,,,
408,100.496420,99.788783,99.039379
415,100.576435,100.503927,100.398187
510,100.097619,99.671429,100.601190


In [29]:
# Unique area codes in the dataset
print(f"Area codes: {df['Area code'].unique()}")

Area codes: [415 408 510]


---
## 12. Feature Engineering <a id='12'></a>

We create new columns by combining existing ones. This can be useful for downstream modeling.

In [30]:
# New feature: Total charge = sum of all individual charges
df['Total charge'] = (
    df['Total day charge']
    + df['Total eve charge']
    + df['Total night charge']
    + df['Total intl charge']
)

print(f"Total charge range: ${df['Total charge'].min():.2f} — ${df['Total charge'].max():.2f}")
df[['Total day charge', 'Total eve charge', 'Total night charge', 'Total intl charge', 'Total charge']].head()

Total charge range: $22.93 — $96.15


,Total day charge,Total eve charge,Total night charge,Total intl charge,Total charge
0,45.07,16.78,11.01,2.70,75.56
1,27.47,16.62,11.45,3.70,59.24
2,41.38,10.30,7.32,3.29,62.29
3,50.90,5.26,8.86,1.78,66.80
4,28.34,12.61,8.41,2.73,52.09


In [31]:
# New feature: Total calls = sum of all call types
df['Total calls'] = (
    df['Total day calls']
    + df['Total eve calls']
    + df['Total night calls']
    + df['Total intl calls']
)

print(f"Total calls range: {df['Total calls'].min()} — {df['Total calls'].max()}")
df[['Total day calls', 'Total eve calls', 'Total night calls', 'Total intl calls', 'Total calls']].head()

Total calls range: 191 — 416


,Total day calls,Total eve calls,Total night calls,Total intl calls,Total calls
0,110,99,91,3,303
1,123,103,103,3,332
2,114,110,104,5,333
3,71,88,89,7,255
4,113,122,121,3,359


In [33]:
df.head() #total calls and total charge is added here

,State,Account length,Area code,International plan,Voice mail plan,Number vmail messages,Total day minutes,Total day calls,Total day charge,Total eve minutes,...,Total night minutes,Total night calls,Total night charge,Total intl minutes,Total intl calls,Total intl charge,Customer service calls,Churn,Total charge,Total calls
0,KS,128,415,False,True,25,265.1,110,45.07,197.4,...,244.7,91,11.01,10.0,3,2.70,1,0,75.56,303
1,OH,107,415,False,True,26,161.6,123,27.47,195.5,...,254.4,103,11.45,13.7,3,3.70,1,0,59.24,332
2,NJ,137,415,False,False,0,243.4,114,41.38,121.2,...,162.6,104,7.32,12.2,5,3.29,0,0,62.29,333
3,OH,84,408,True,False,0,299.4,71,50.90,61.9,...,196.9,89,8.86,6.6,7,1.78,2,0,66.80,255
4,OK,75,415,True,False,0,166.7,113,28.34,148.3,...,186.9,121,8.41,10.1,3,2.73,3,0,52.09,359


---
## ✅ Summary of Key Findings

| Insight | Detail |
|---|---|
| **Churn rate** | ~14.5% — high risk for the business |
| **Class imbalance** | 85% loyal vs 15% churned |
| **Day minutes** | Churned users average ~207 min/day vs ~175 for loyal users |
| **International plan & Churn** | Churned customers are ~4x more likely to have an international plan (28.36% vs 6.53%) |
| **International plan & Call Duration** | Customers with international plans generally have slightly higher call durations across all times of day. |
| **Customer service calls** | Churned users averaged ~2.2 calls vs ~1.4 for loyal users |
| **No missing data** | Dataset is clean — ready for modeling |